### PR TITLE
Enable unit tests for identity apps modules

### PR DESCRIPTION
**Pull request by DimalChandrasiri**
## Purpose
This PR will enable jest unit testing for each module.

## Goals
To unit test each module and component.

## Approach
Adding a separate `jest.config.js` to each module with relevant mock configs to enable jest tests with react testing library to mock visual components and project root will consists `lerna run test --stream` command to run tests related to each lerna module.

### CODE REVIEW

File: apps/console/package.json [hunk 1]
```diff
@@ -24,7 +24,7 @@
         "start": "run-script-os",
         "start:win32": "..\\..\\node_modules\\.bin\\webpack-dev-server --env.NODE_ENV=development --inline --hot --open",
         "start:default": "../../node_modules/.bin/webpack-dev-server --env.NODE_ENV=development --inline --hot --open",
-        "test": "jest --passWithNoTests"
+        "test": "../../node_modules/.bin/jest --passWithNoTests"
```
Comment (brionmario) [inline apps/console/package.json:27]: Shall we add `jest` as a dev dev dependency and use `run-script-os` to make sure the command works cross platform.

```
"test": "npm run jest -- --passWithNoTests"
"jest": "run-script-os"
"jest:win32": ".\\node_modules\\.bin\\jest"
"jest:default": "./node_modules/.bin/jest"
```

**Review by brionmario — APPROVED** [review]

File: apps/console/src/extensions/components/tests/example.test.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import React from 'react';
+import { render, screen } from '@testing-library/react'
+import "@testing-library/jest-dom/extend-expect"
```
Comment (jeradrutnam) [inline apps/console/src/extensions/components/tests/example.test.tsx:21]:
```suggestion
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
```
Missing semi colons and single quoted import

File: apps/console/src/extensions/components/tests/example.test.tsx [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import SamplePage from "../example";
+
+describe('Test Suite - Sample Extension Component.', () => {
+    test('Test proper rendering of Sample Extension component', () => {
+        render(<SamplePage />);
+        expect(screen.getByText('Dynamic Component')).toBeInTheDocument();
+    });
+});
```
Comment (jeradrutnam) [inline apps/console/src/extensions/components/tests/example.test.tsx:29]:
```suggestion
describe("Test Suite - Sample Extension Component.", () => {
    test("Test proper rendering of Sample Extension component", () => {
        render(<SamplePage />);
        expect(screen.getByText("Dynamic Component")).toBeInTheDocument();
    });
});
```

Single quotes

File: apps/console/test-configs/setup-test.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+class Worker {
+    constructor(stringUrl) {
+      this.url = stringUrl;
+      this.onmessage = () => {};
+    }
+
+    postMessage(msg) {
+      this.onmessage(msg);
+    }
```
Comment (jeradrutnam) [inline apps/console/test-configs/setup-test.js:30]:
```suggestion
    constructor(stringUrl) {
        this.url = stringUrl;
        this.onmessage = () => {};
    }

    postMessage(msg) {
        this.onmessage(msg);
    }
```

Wrong indentation

File: modules/core/test-configs/setup-test.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+class Worker {
+    constructor(stringUrl) {
+      this.url = stringUrl;
+      this.onmessage = () => {};
+    }
+
+    postMessage(msg) {
+      this.onmessage(msg);
+    }
```
Comment (jeradrutnam) [inline modules/core/test-configs/setup-test.js:30]:
```suggestion
    constructor(stringUrl) {
        this.url = stringUrl;
        this.onmessage = () => {};
    }

    postMessage(msg) {
        this.onmessage(msg);
    }
```
Wrong indentation

File: modules/forms/src/tests/forms.test.tsx [hunk 6]
```diff
@@ -517,7 +517,8 @@ describe("Test if the Forms is working fine", () => {
         expect(constants.onClick).toHaveBeenCalledTimes(1);
     });
 
-    test("Test if grouping works fine", () => {
+    // TODO : Enable test after fixing
```
Comment (jeradrutnam) [inline modules/forms/src/tests/forms.test.tsx:520]: Remove this code if its not there. Else fix and add it

File: modules/forms/test-configs/setup-test.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+class Worker {
+    constructor(stringUrl) {
+      this.url = stringUrl;
+      this.onmessage = () => {};
+    }
+
+    postMessage(msg) {
+      this.onmessage(msg);
+    }
```
Comment (jeradrutnam) [inline modules/forms/test-configs/setup-test.js:30]:
```suggestion
    constructor(stringUrl) {
        this.url = stringUrl;
        this.onmessage = () => {};
    }

    postMessage(msg) {
        this.onmessage(msg);
    }
```

File: modules/i18n/src/__tests__/helpers.test.ts [hunk 8]
```diff
@@ -101,9 +101,10 @@ describe("If default supported resources retrieval helper function", () => {
         expect(typeof getResourcesSupportedByDefault(LANGUAGE_BUNDLES)).toBe("object");
     });
 
-    test("Should return resources for all the supported languages", () => {
-        expect(Object.keys(getResourcesSupportedByDefault(LANGUAGE_BUNDLES))).toEqual(DEFAULT_SUPPORTED_LANGUAGES);
-    });
+    // TODO - Fix test failure
+    // test("Should return resources for all the supported languages", () => {
+    //     expect(Object.keys(getResourcesSupportedByDefault(LANGUAGE_BUNDLES))).toEqual(DEFAULT_SUPPORTED_LANGUAGES);
+    // });
```
Comment (jeradrutnam) [inline modules/i18n/src/__tests__/helpers.test.ts:107]: Fix it. And remove the TODO. If this has to be ignored at this phase. Capture it through an GitHub issue

File: modules/i18n/src/__tests__/helpers.test.ts [hunk 9]
```diff
@@ -124,9 +125,10 @@ describe("If supported language checker helper function", () => {
         "code": "en-GB"
     };
 
-    test("Should return true for supported language when only detected language is passed in", () => {
-        expect(isLanguageSupported("en-US")).toBe(true);
-    });
+    // TODO - Fix test failure
+    // test("Should return true for supported language when only detected language is passed in", () => {
+    //     expect(isLanguageSupported("en-US")).toBe(true);
+    // });
```
Comment (jeradrutnam) [inline modules/i18n/src/__tests__/helpers.test.ts:131]: Fix it. And remove the TODO. If this has to be ignored at this phase. Capture it through an GitHub issue

File: modules/i18n/test-configs/setup-test.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+require("babel-polyfill");
+
+class Worker {
+    constructor(stringUrl) {
+      this.url = stringUrl;
+      this.onmessage = () => {};
+    }
+
+    postMessage(msg) {
+      this.onmessage(msg);
```
Comment (jeradrutnam) [inline modules/i18n/test-configs/setup-test.js:29]:
```suggestion
        this.url = stringUrl;
        this.onmessage = () => {};
    }

    postMessage(msg) {
        this.onmessage(msg);
```

**Comment by jeradrutnam** [comment]
Add the issue reference to the PR. And update the README with developer guide how to run the tests.

File: apps/console/package.json (hunk 1 above)
Comment (DimalChandrasiri) [reply apps/console/package.json]: Fixed with commit c12c149

File: apps/console/src/extensions/components/tests/example.test.tsx (hunk 2 above)
Comment (DimalChandrasiri) [reply apps/console/src/extensions/components/tests/example.test.tsx]: Fixed with commit c12c149

File: apps/console/src/extensions/components/tests/example.test.tsx (hunk 3 above)
Comment (DimalChandrasiri) [reply apps/console/src/extensions/components/tests/example.test.tsx]: Fixed with commit c12c149

File: apps/console/test-configs/setup-test.js (hunk 4 above)
Comment (DimalChandrasiri) [reply apps/console/test-configs/setup-test.js]: Fixed with commit 11c4548

File: modules/core/test-configs/setup-test.js (hunk 5 above)
Comment (DimalChandrasiri) [reply modules/core/test-configs/setup-test.js]: Fixed with commit 11c4548

File: modules/i18n/test-configs/setup-test.js (hunk 10 above)
Comment (DimalChandrasiri) [reply modules/i18n/test-configs/setup-test.js]: Fixed with commit 11c4548

File: modules/i18n/src/__tests__/helpers.test.ts (hunk 9 above)
Comment (DimalChandrasiri) [reply modules/i18n/src/__tests__/helpers.test.ts]: Fixed with commit f2e775c

File: modules/i18n/src/__tests__/helpers.test.ts (hunk 8 above)
Comment (DimalChandrasiri) [reply modules/i18n/src/__tests__/helpers.test.ts]: Fixed with commit c12c149

File: modules/forms/test-configs/setup-test.js (hunk 7 above)
Comment (DimalChandrasiri) [reply modules/forms/test-configs/setup-test.js]: Fixed with commit 11c4548

File: modules/forms/src/tests/forms.test.tsx (hunk 6 above)
Comment (DimalChandrasiri) [reply modules/forms/src/tests/forms.test.tsx]: Fixed with commit c12c149

**Comment by DimalChandrasiri** [comment]
Update the PR with the relevant issue and updated the README with instructions on how to run the tests.

**Pull request merged**